### ISSUE TITLE
[DX] Drop support for php 5.6 - or change our policy re smallest version shipped with supported major linux distros

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5920]
This is a follow-up issue to #3992

Backdrop core currently supports php versions from 5.6 and up (see [the announcement](https://backdropcms.org/news/backdrop-cms-drops-support-for-php-versions-prior-to-56)), but there are certain cases where this becomes problematic. I am starting a register of such cases here, along with linked example issues in the core queue:
- type hinting of types other than objects and arrays: #5890
- ???

PHP 5.6 has been EoL since 31 Dec 2018 (see https://www.php.net/eol.php), which means no official security fixes. However, LTS (long-term support) versions of various linux distributions being used by popular hosting providers have been backporting security fixes and major bug fixes to 5.6, so we couldn't exactly call 5.6 EoL and be done with it. Also, Zend seems to be providing a paid LTS service for any PHP version after and including 5.6: https://www.zend.com/services/php-long-term-support

---

Some stats (at the time of writing - Jan 2023). Let's try to keep these up to date as this issue here is being considered:
- https://packagist.org/php-statistics (showing roughly ~1% combined usage of all versions of php 7.0 and below in Jan 2023)
  ![image](https://user-images.githubusercontent.com/2423362/211652196-d906dfcd-707a-4f16-a971-25cb91426979.png)
- https://w3techs.com/technologies/details/pl-php/all/all (apparently still not below 20% globally as of Jan 2023)
  ![image](https://user-images.githubusercontent.com/2423362/211652574-29360323-325b-4493-b9c9-781b90feefd2.png)
- https://wordpress.org/about/stats
  <img src="https://user-images.githubusercontent.com/2423362/211653004-5ac17d79-f5e2-4095-a703-bd26ec0b524a.png" width=40%>
- our own telemetry stats: https://backdropcms.org/project/backdrop/telemetry (showing only 2% of roughly 325 sites on php 5.6)

**Comment by klonos** [comment]
...I have tentatively set this task for the Backdrop 2.x milestone, which won't come before Jan 2025. We may decide to reconsider soon er than that though.

Does D7 EoL affect this decision here at all? If so, then how?

**Comment by herbdool** [comment]
We are still following the policy of supporting the lowest version still supported by a Linux distro. See https://github.com/backdrop/backdrop-issues/issues/3992#issuecomment-949252733.

So that looks like 7.2?

**Comment by klonos** [comment]
Yes @herbdool, that'd be right I believe 👍🏼 

FTR, I'm adding here the table that @quicksketch had compiled back then:

| Distro      | Release Date | EOL Date   | PHP Version |
| ----------- | ------------ | ---------- | ----------- |
| CentOS 6    | 2011-07-10   | 2020-11-30 | 5.3         |
| CentOS 7    | 2014-07-07   | 2024-06-30 | 5.4         |
| CentOS 8    | 2019-09-24   | 2021-12-31* | 7.3         |
| RHEL 6      | 2010-11-10   | 2020-11-30 | 5.3         |
| RHEL 7      | 2014-06-10   | 2021-08-30 | 5.4         |
| RHEL 8      | 2019-05-07   | 2029       | 7.3         |
| Debian 8    | 2015-04-25   | 2020-06-30 | 5.6         |
| Debian 9    | 2017-06-17   | 2022-06-30 | 7.0         |
| Debian 10   | 2019-07-06   | 2024-06    | 7.2         |
| Debian 11   | 2021-08-14   | 2026-06    | 7.4         |

**Comment by klonos** [comment]
...and also FTR, here's what currently being reported for the various php versions in our telemetry:

Value | Count | Percentage
-- | -- | --
7.1 | 1 | 0%
5.6 | 7 | 2%
7.0 | 14 | 4%
7.3 | 26 | 8%
8.1 | 33 | 10%
8.0 | 37 | 11%
7.2 | 45 | 13%
7.4 | 173 | 51%

**Comment by yorkshire-pudding** [comment]
> * our own telemetry stats: https://backdropcms.org/project/backdrop/telemetry (showing only 2% of roughly 325 sites on php 5.6)

I wonder whether it is possible from the raw telemetry stats (or whether it would need to be added) to know what version someone is on; that would give some kind of indication as to the active management of the site. I guess we do know that they're on a version after telemetry was added.

It would also be interesting to understand why people are still on 5.6. My guess is inertia; nothings broke so nothing to fix and host not encouraging then to move.

**Comment by jenlampton** [comment]
I'm also in favor of waiting on removing support for PHP 5.6 until backdrop 2.0. (Perhaps at that point, our requirements can be updated to PHP 8?)

I'm concerned that existing Backdrop sites aren't going to be upgrading the PHP versions. Even if PHP 5 isn't supported, there will still be web hosts offering patched versions, and if there is other old software on the same instance it could be too much of a risk to update the PHP version just to support the Backdrop site.

**Comment by yorkshire-pudding** [comment]
My hosts supports from 5.2 to 8.2 using cPanel, so I hope it is a minority.
![image](https://user-images.githubusercontent.com/20357797/212283849-b4efe67e-5a8f-49ea-ab26-10a494047411.png)

**Comment by klonos** [comment]
Yeah, lets keep this as a placeholder. I'll try my best to be keeping the stats/metrics up to date from time to time, and as suggested during this week's dev meeting, perhaps once we have a bigger percentage of sites with Telemetry enabled, perhaps we can make a better-informed decision then.

**Comment by avpaderno** [comment]
Was not support for PHP 5.6 already dropped? `system_requirements()` uses the following code.

```php
// Minimum recommended PHP version can be incremented when deprecating older
// versions of PHP but while they are still supported. This does not throw
// a warning or error on updates to keep upgrade tests compatible.
$minimum_recommended_version = '7.1.0';
$minimum_warning = $t('Your PHP installation is too old. Backdrop requires at least PHP %version.', array('%version' => $minimum_recommended_version));
if (version_compare($phpversion, $minimum_recommended_version) < 0) {
  $requirements['php']['description'] = $minimum_warning . ' ' . l(t('Read about the minimum version increase'), 'https://backdropcms.org/news/backdrop-cms-drops-support-for-php-versions-prior-to-71') . '.';
  $requirements['php']['severity'] = $phase === 'runtime' ? REQUIREMENT_ERROR : REQUIREMENT_INFO;
}

// If PHP does not meet the bare minimum, remaining checks are unnecessary.
if (version_compare($phpversion, BACKDROP_MINIMUM_PHP) < 0) {
  $requirements['php']['description'] = $minimum_warning;
  $requirements['php']['severity'] = REQUIREMENT_ERROR;
  return $requirements;
}
```

At least, the intention seems that.

**Comment by avpaderno** [comment]
(There is [Backdrop CMS drops support for PHP versions prior to 7.1](https://backdropcms.org/news/backdrop-cms-drops-support-for-php-versions-prior-to-71), which says PHP 7.1 is the minimum required version.)

**Comment by herbdool** [comment]
Yes, we can close this.
